nes-sinks/src/SinkCatalog.cpp

```diff
@@ -0,0 +1,88 @@
+/*
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        https://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+#include <Sinks/SinkCatalog.hpp>
+
+#include <algorithm>
+#include <cctype>
+#include <memory>
+#include <mutex>
+#include <optional>
+#include <string>
+#include <string_view>
+#include <unordered_map>
+#include <utility>
+#include <DataTypes/Schema.hpp>
+#include <Sinks/SinkDescriptor.hpp>
+#include <Util/Logger/Logger.hpp>
+
+namespace NES
+{
+std::optional<NES::Sinks::SinkDescriptor> NES::SinkCatalog::addSinkDescriptor(
+    std::string sinkName, const Schema& schema, const std::string_view sinkType, std::unordered_map<std::string, std::string> config)
+{
+    if (std::ranges::all_of(sinkName, [](const char character) { return std::isdigit(character); }))
+    {
+        /// Reserve number names for the future where we might want to have inlined sink descriptor declarations in queries again, which requires autogenerating some identifier.
+        NES_DEBUG("Sink name {} is invalid, only-digit names are reserved for autogenerated names", sinkName);
+        return std::nullopt;
+    }
+    auto schemaCopy = std::make_shared<Schema>(schema);
+    auto descriptorConfigOpt = Sinks::SinkDescriptor::validateAndFormatConfig(sinkType, std::move(config));
+    if (not descriptorConfigOpt.has_value())
+    {
+        return std::nullopt;
+    }
+
+    const std::unique_lock lock(catalogMutex);
+    auto sinkDescriptor = Sinks::SinkDescriptor(sinkName, std::move(schemaCopy), sinkType, std::move(descriptorConfigOpt.value()));
+    sinks.emplace(std::move(sinkName), sinkDescriptor);
+    return sinkDescriptor;
+}
+
+std::optional<NES::Sinks::SinkDescriptor> NES::SinkCatalog::getSinkDescriptor(const std::string& sinkName) const
+{
+    const std::unique_lock lock(catalogMutex);
+    const auto sinkDescriptorOpt = sinks.find(std::string{sinkName});
+    if (sinkDescriptorOpt == sinks.end())
+    {
+        return std::nullopt;
+    }
+    return sinkDescriptorOpt->second;
+}
+
+bool NES::SinkCatalog::removeSinkDescriptor(const std::string& sinkName)
+{
+    const std::unique_lock lock(catalogMutex);
+    return sinks.erase(sinkName) == 1;
+}
+
+bool NES::SinkCatalog::removeSinkDescriptor(const Sinks::SinkDescriptor& sinkDescriptor)
```

Comment:
You pass the sink descriptor itself here and the name to the `contains` method. Keep it consistent with respect to what the actual use is (i.e., do we want to pass the name of the descriptor here).